docs/source/format/Columnar.rst

```diff
@@ -765,6 +765,68 @@ application.
 We discuss dictionary encoding as it relates to serialization further
 below.
 
+.. _run-length-encoded-layout:
+
+Run-Length Encoded Layout
+-------------------------
+
+Run-Length is a data representation that represents data as sequences of the
+same value, called runs. Each run is represented as a value, and an integer
+describing how often this value is repeated.
+
+Any array can be run-length encoded. A run-length encoded array has no buffers
+by itself, but has two child arrays. The first one holds a signed integer
+called a "run end" for each run. The run ends array can hold either 16, 32, or
+64-bit integers. The actual values of each run are held
+the second child array.
+
+The values in the first child array represent the length of each run. They do
+not hold the length of the respective run directly, but the accumulated length
+of all runs from the first to the current one, i.e. the logical index where the
+current run ends. This allows relatively efficient random access from a logical
+index using binary search. The length of an individual run can be determined by
+subtracting two adjacent values.
+
+A run must have have a length of at least 1. This means the values in the
+run ends array all positive and in strictly ascending order. A run end cannot be
+null.
+
+As an example, you could have the following data: ::
+
+    type: Float32
+    [1.0, 1.0, 1.0, 1.0, null, null, 2.0]
+
+In Run-length-encoded form, this could appear as:
+
+::
+
+    * Length: 7, Null count: 2
```

Comment:
This formulation of null count is a little surprising, it implies that when computing the parent ArrayData from a set of children, it must iterate the null mask in conjunction with the run ends. This is also inconsistent with how null counts work for other nested types

Reply:
You're right, it should actually say 0, since the array has no validity bitmap

(I tried to reply via the reply button in the ml web ui but that does not seem to have come through)

Reply:
> since the array has no validity bitmap

Is this a general restriction, or is it just in this case there is no validity mask?

Reply:
Yes, this is a general restriction (at least I think so, and that's how the prototype code works). The idea so far is that Null is just one of the possible values for a run.

The is somewhat consistent with how union types work.

(if we were to allow the RLE array parent to have an additional null mask, the null count field would represent that - there seems to be a generall assumption in Arrow code that a non-zero (or array length for the NULL) null count means the presence of the standard null mask)

Reply:
> this is a general restriction

We should probably document that if so, personally it seems a shame as it complicates things like nullif kernels which now have to explicitly handle the case of RLE data, whereas normally they don't need concern themselves with anything but the array data.

> union types work

Yeah, union types are a bit of a special snowflake though :laughing:

Reply:
it is actually documented by saying there are no buffers, but that's a bit indirect, so guess mentioning it explicitly won't hurt

Reply:
I submitted a PR for this here, please take a look:
[https://github.com/apache/arrow/pull/33831](https://github.com/apache/arrow/pull/33831)